main.tf

```diff
@@ -1,236 +1,428 @@
-
-  api_id    = aws_apigatewayv2_api.this[0].id
-  route_key = each.key
-
-  api_key_required                    = try(each.value.api_key_required, null)
-  authorization_scopes                = try(split(",", each.value.authorization_scopes), null)
-  authorization_type                  = try(each.value.authorization_type, "NONE")
-  authorizer_id                       = try(aws_apigatewayv2_authorizer.this[each.value.authorizer_key].id, each.value.authorizer_id, null)
-  model_selection_expression          = try(each.value.model_selection_expression, null)
-  operation_name                      = try(each.value.operation_name, null)
-  route_response_selection_expression = try(each.value.route_response_selection_expression, null)
-  target                              = "integrations/${aws_apigatewayv2_integration.this[each.key].id}"
+  api_id         = aws_apigatewayv2_api.this[0].id
+  integration_id = aws_apigatewayv2_integration.this[each.key].id
 
-  # Have been added to the docs. But is WEBSOCKET only(not yet supported)
-  # request_models  = try(each.value.request_models, null)
+  content_handling_strategy     = each.value.response.content_handling_strategy
+  integration_response_key      = each.value.response.integration_response_key
+  response_templates            = each.value.response.response_templates
+  template_selection_expression = each.value.response.template_selection_expression
 }
 
-resource "aws_apigatewayv2_integration" "this" {
-  for_each = { for k, v in var.integrations : k => v if var.create && var.create_routes_and_integrations }
+################################################################################
+# Stage
+################################################################################
+
+locals {
+  create_stage = var.create && var.create_stage
+
+  default_log_format = jsonencode({
+    context = {
+      domainName              = "$context.domainName"
+      integrationErrorMessage = "$context.integrationErrorMessage"
+      protocol                = "$context.protocol"
+      requestId               = "$context.requestId"
+      requestTime             = "$context.requestTime"
+      responseLength          = "$context.responseLength"
+      routeKey                = "$context.routeKey"
+      stage                   = "$context.stage"
+      status                  = "$context.status"
+      error = {
+        message      = "$context.error.message"
+        responseType = "$context.error.responseType"
+      }
+      identity = {
+        sourceIP = "$context.identity.sourceIp"
+      }
+      integration = {
+        error             = "$context.integration.error"
+        integrationStatus = "$context.integration.integrationStatus"
+      }
+    }
+  })
+}
 
-  api_id      = aws_apigatewayv2_api.this[0].id
-  description = try(each.value.description, null)
+resource "aws_apigatewayv2_stage" "this" {
+  count = local.create_stage ? 1 : 0
```

Comment:
I wonder how often/reasonably it is to create multiple "stages" in an API Gateway?

If I remember correctly, "$default" was a special stage name. Do we need to have support to manage more stages?

Reply:
> I wonder how often/reasonably it is to create multiple "stages" in an API Gateway?

Asking around internally on this, it is possible to have multiple stages per API gateway, however it is not recommended to use stages for separation of environments (i.e. - `dev` stage, `qa` or `uat` stage, and then `prod` or `$default` stage for production). We've only had one issue raised for this https://github.com/terraform-aws-modules/terraform-aws-apigateway-v2/issues/94

I also took a look at what it might look like if we were to entertain multiple stages per API Gateway and I don't think that is feasible unless we really separate things which looks like that would hurt the usefulness of the module. If some of the resources were split up more finely, such as the `route_settings` of the `aws_apigatewayv2_stage` , or if those were moved over to the `aws_apigatewayv2_route` resource, then it might be more feasible. But right now, the majority of the resources are to tightly intertwined. So between the lack of high demand for multiple stages on an API Gateway, and the challenge with trying to support that, I would lean towards not supporting that at this time